### PR TITLE
TrustStore support for OIDC and OIDCClient

### LINKED ISSUE #18002: Custom CA for OIDC connection
## Description

When the connection to OIDC auth server is encrypted using TLS the certificate must be trusted by the system, or we have to disable verification completely using `quarkus.oidc.tls.verification=none`.
This is inconvenient when the certificate is provided in Openshift by Service CA that gets mounted as a configmap - the certificate mounted into the image cannot be made trusted system-wide (without elevated privileges) but we cannot set it in Quarkus, neither as PEM certificate nor as a Java Keystore.
In addition to that mutual TLS is not possible at all because we cannot set client certificate.

## Implementation ideas

We should follow the same approach as with Resteasy client; properties like `quarkus.oidc.tls.trustStore` should be added.

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #18002.

This draft PR adds support for configuring OIDC, OIDC client and also Keycloak authorization with the trust store file/password - property names are the same as in Vert.x HTTP, we can add more related properties as needed (alias, etc). 

CC @rvansa

### CODE REVIEW

File: test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java
Comment (sberyozkin) [inline test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java:74]: @pedroigor is it the right way to mount a key ? I've read in in a rather old Stackoverflow that this is how it should be done, but I'm not sure it is still the case, the test does not work

Comment (sberyozkin) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java:74]: @pedroigor I have not seen in the latest docs any references to a private key having to be mounted to `/etc/x509/http/tls.key`, it is only mentioned in the non up to date docs. I wonder what is the current approach is - mounting a private key should lead to WildFly converting it to a keystore, but perhaps now an actual keystore file should be mapped ?

**Comment by sberyozkin** [comment]
@rvansa Would you like to try this PR against your deployment ? If we'll have problems verifying it with the test then I guess we can just merge it without it.

**Comment by rvansa** [comment]
@sberyozkin Hi, thanks a lot for the fix. I'll try to verify it soon but I have pretty full hands this week and the next one I won't be available, so have some patience please.

**Comment by sberyozkin** [comment]
@rvansa Sounds good :-), I'll be off next week myself, thanks

**Comment by famod** [comment]
I'll try to put that on my list as well since we also have to use TLS to talk to our Keycloak container and in dev env we also use custom certs.

File: test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java
Comment (pedroigor) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java:74]: Perhaps it should be `/etc/x509/https/tls.key`?

Comment (sberyozkin) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java]: @pedroigor thanks, the docs at https://hub.docker.com/r/jboss/keycloak/ do refer to `/etc/x509/https`.

I'm pretty sure some progress has been made as a result since now I'm seeing now in the logs:

```
Creating HTTPS keystore via OpenShift's service serving x509 certificate secrets..
HTTPS keystore not created at: /opt/jboss/keycloak/standalone/configuration/keystores/https-keystore.jks (check permissions?)
```

I've checked, this error is not logged if the path is wrong (`http` in `/etc/x509/http/tls.key`).

The section at https://hub.docker.com/r/jboss/keycloak/ points out at the possible problem.

The permissions for both `tls.crt` and `tls.key` are `r` for the world on my system.
Host mounting them with `withClasspathResourceMapping` or copying to the docker (`.withCopyFileToContainer`) makes no difference though.... 
But it is nearly there, we just need to see how it is done in some working deployments

Comment (sberyozkin) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/server/KeycloakTestResourceLifecycleManager.java]: @pedroigor @rvansa the example [here](https://stackoverflow.com/questions/49859066/keycloak-docker-https-required) shows how these files are mapped with the specific user/group ids and also the mode. `TestContainers` does now allow to set it when mounting the files.
I've also seen somewhere setting the mode to `655` on the local system before mounting the files should work - but I could not make it work.

**Comment by famod** [comment]
I haven't yet gotten around to testing this but I have it on my list for this week. Would be great to have this end up in Quarlus 2.1

Btw, `main` is now on vert.x 4.1.1 which offers `aliasPassword` (and 4.1 offers `alias`). #18387 is adding `aliasPassword` to quarkus-vertx-http.

**Comment by sberyozkin** [comment]
@famod Thanks, I believe you meant `trustStoreKeyAlias`, I'll add this property as well

**Comment by famod** [comment]
I was actually thinking of the next step that would be client cert auth (for mutual TLS).

**Comment by sberyozkin** [comment]
@famod 

> I was actually thinking of the next step that would be client cert auth (for mutual TLS).

OK

**Comment by famod** [comment]
@sberyozkin Sorry, didn't have the time yet to look at it. Do you think its still realistic to get this into 2.1?

**Comment by sberyozkin** [comment]
@famod, np at all - we need at least a single confirmation this PR actually helps - I can try to experiment as well, haven't had time yet; so not sure if it can make it, I believe it would be safe to backport though

**Comment by sberyozkin** [comment]
@famod Hi, I've managed to verify this code against a standalone Keycloak instance, by following [Server installation Guide](https://www.keycloak.org/docs/latest/server_installation/#_setting_up_ssl) and also updating a custom realm configuration to require SSL for all the requests as described [here](https://www.keycloak.org/docs/latest/server_admin/#_ssl_modes). I copied the JKS file used in this PR, here are the relevant fragment:

[source,xml]
----
        <subsystem xmlns="urn:jboss:domain:undertow:12.0" default-server="default-server" default-virtual-host="default-host" default-servlet-container="default" default-security-domain="other" statistics-enabled="${wildfly.undertow.statistics-enabled:${wildfly.statistics-enabled:false}}">
            <buffer-cache name="default"/>
            <server name="default-server">

                <https-listener name="https" socket-binding="https" security-realm="UndertowRealm"/>

        </subsystem>
----

and

[source,xml]
----
<management>
        <security-realms>
<security-realm name="UndertowRealm">
    <server-identities>
        <ssl>
            <keystore path="keycloak.jks" relative-to="jboss.server.config.dir" keystore-password="secret" />
        </ssl>
    </server-identities>
</security-realm>
        </security-realms>
    </management>
----

I've also has to add a new enum, `CERTIFICATE_VALIDATION` to support a case where only the certificate validation but no hostname verification is required. I also imported the existing realm from `quarkus-quickstarts/security-openid-connect-quickstart` 

Finally I've tweaked `BearerTokenAuthorizationTest`:
1. Commented //@QuarkusTestResource(KeycloakTestResourceLifecycleManager.class)
2. Copied `getAccessToken` from `KeycloakTestResourceLifecycleManager` to `BearerTokenAuthorizationTest` and updated to set the truststore path and secret exactly as it is done in `KeycloakTestResourceLifecycleManager` (and also use `backend-client` client id which is created from the existing realm)
3. Updated `application.properties` to have `auth-server-url` to point to `https://localhost:8543/auth/realms/quarkus`  
4. Removed a ref to the `email` claim from `application.properties` as it is not avail in that existing realm
5. Finally run a slightly updated test `BearerTokenAuthorizationTest#testSecureAccessSuccessCustomPrincipal`

```
@Test
    public void testSecureAccessSuccessCustomPrincipal() {
            RestAssured.given().auth().oauth2(getLocalAccessToken("alice"))
                    .when().get("/api/users/me")
                    .then()
                    .statusCode(200)
                    .body("userName", equalTo("alice"));
    }
```

and it worked.

I've temporarily disabled though the test resource factory code which sets up the trust store, I'll open an issue afterwards to try and get the test running, I've found another resource:
https://strimzi.io/blog/2019/10/25/kafka-authentication-using-oauth-2.0/

which suggests that the test factory does exactly the same thing by mounting the key and cert, but something just does not click - all these docs refer to the old image, so it needs some support from the KC guys.

Have a look please and I'll merge once you approve

**Comment by famod** [comment]
I have just started testing this PR with my current Quarkus project (that incorporates KC).

**Comment by famod** [comment]
@sberyozkin Hum...it seems like I'm hitting a wall here as I keep getting:
```
io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://idp:8443/auth/realms/egbr' URL
[...]
 Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```
I seems as if the entry in my store is not taken into account. It does find the store though because settings some invalid value for `quarkus.oidc.tls.trust-store-file` does yield an exception (as expected).

I even tried setting the alias.

**Comment by famod** [comment]
Nevermind, the store was missing our custom CA cert (it only contained the actual server cert). I will conclude my review tomorrow.
But looking good so far! 🎉

**Review by famod — APPROVED** [review]
Works in my project/setup and code LGTM.

I'll trust you on that test factory part @sberyozkin 🙂, didn't have a look at it myself.

I'll mark this for backport and see how it goes.

**Comment by famod** [comment]
> Have a look please and I'll merge once you approve

I read that a bit wrong and merged it already but I don't think it matters who merges in the end. 😉

**Comment by sberyozkin** [comment]
@famod cool, thanks for checking it (I just came here to mention that `8543` port was used because I started with a `100` port offset :-) )

**Comment by sberyozkin** [comment]
@famod in the next iteration we can wire in a mutual tls authentication, will have a look later on

**Pull request merged**
